app/permissions.py

```diff
@@ -0,0 +1,52 @@
+from functools import wraps
+from typing import Any, Callable
+
+from django.contrib import messages
+from django.contrib.auth import logout
+from django.http import HttpRequest, HttpResponseRedirect
+from django.shortcuts import redirect, reverse
+
+from app.middleware import refresh_token
+
+
+def require_coc(func: Callable) -> Any:
+    """Enforce code of conduct acceptance."""
+
+    @wraps(func)
+    def inner_func(request: HttpRequest, *args: list, **kwargs: dict) -> Any:
+        if not request.user.accepted_coc:
+            return redirect("accept_coc")
+        return func(request, *args, **kwargs)
+
+    return inner_func
```

Comment:
Where does this `inner_func` come from? From the annotation?

Reply:
It's a decorator, so we're returning the function that's declared right above this.